Send Signer results over a channel

xoloki opened the pull request:
### Description

Right now the only way to find out what a `Signer` has done is to stop it and get the cached results.  

Add another channel which lets `Signer` send results as they come in.

### Applicable issues
- Closes https://github.com/stacks-network/stacks-blockchain/issues/3916

### Additional info (benefits, drawbacks, caveats)

### Checklist
- [ ] Test coverage for new or modified code paths
- [ ] Changelog is updated
- [ ] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)
- [ ] New clarity functions have corresponding PR in `clarity-benchmarking` repo
- [ ] New integration test(s) added to `bitcoin-tests.yml`

xoloki committed:
use a channel to send the results from the SignerRunLoop to whoever is interested; remove sleep calls in test since we know when it's done

jcnelson commented on libsigner/src/session.rs:
```diff
@@ -96,9 +96,9 @@ impl StackerDBSession {
-        if self.sock.is_none() {
-            self.connect_or_reconnect()?;
-        }
+        // TODO: fix this so we can use persistent connection
```
I strongly suspect that the node is closing the socket after it dumps the event, but not on a predictable schedule (see https://github.com/stacks-network/stacks-blockchain/blob/master/testnet/stacks-node/src/event_dispatcher.rs#L126).  The socket gets consumed by `async_h1`, which presumably drops it and closes it at some future date.  A next-step on this issue would be to update the node to either (1) always close the socket _explicitly_ once the payload has been sent, or (2) require `async_h1` to keep it open so the TCP window can at least be reused.  I don't care which one happens, but it would at least let us address this comment.

jferrant replied:
I have created a ticket already for this https://github.com/stacks-network/stacks-blockchain/issues/3922 . I will link to this comment for additional context :D

jcnelson commented on stacks-signer/src/config.rs:
```diff
@@ -38,25 +43,70 @@ pub enum ConfigError {
+    pub stacks_private_key: String,
+    /// The network to use. One of "mainnet" or "testnet".
+    pub network: Network,
+    // TODO: Optionally retrieve the signers from the pox contract
```
Doesn't have to be done in this PR ofc, but it's not optional ;)  The default behavior will be to obtain the signers' public keys from `pox-4` (or, at the user's discretion, a conforming smart contract that could be used to mock `pox-4` for testing).

jcnelson replied:
Just commenting here so it gets on y'all's radar.

jferrant replied:
Only reason I have this marked optional is because I could see for testing purposes that we should be able to provide this information in a config rather than just always from the pox contract.

jcnelson commented on stacks-signer/src/stacks_client.rs:
```diff
@@ -0,0 +1,114 @@
+}
+
+/// Helper function to determine the slot ID for the provided stacker-db writer id and the message type
+fn slot_id(id: u32, message: &MessageTypes) -> u32 {
```
Right now, stackerdb only permits up to 4096 slots per replica.  If we have 2000 signers (one signer per reward slot), this would mean each signer gets two slots.

The reason for the 4096 cap is because nodes must exchange slot version vectors with one another to detect the availability of new chunks, and the size of the message is basically 4x the number of slots (each version is 4 bytes).  The cap could be raised, but I'm not sure yet that its wise to make it an order of magnitude bigger.  I'd want to do some performance tests to see if nodes regularly exchanging 160kb chunks of data with up to 32 neighbors on each write is sustainable (that's just over 5 MB write amplification per slot push).

jcnelson replied:
If it turns out that the 4096 cap is the upper bound, would it be feasible to adapt this code later on so that each signer gets a single slot?  If we can swing this, then the task of setting up the next set of stacker signers at a reward cycle boundary is made easier:  the first 2048 slots would belong to the current signers, and the next 2048 slots would belong to the next signers coming online.

xoloki replied:
As we discussed in the first Signer PR, we only need one set of slots per Signer, not per reward slot.  So 2048 is more than enough for our use case with ~100 Signers.

jcnelson commented on stacks-signer/src/tests/contracts/signers-dkg.clar:
```diff
@@ -0,0 +1,37 @@
+
+;; traits
+;;
+(impl-trait 'ST1NXBK3K5YYMD6FD41MVNP3JS1GABZ8TRVX023PT.sip-010-trait-ft-standard.sip-010-trait)
```
Is there a reason this code needs to implement SIP-010?  If so, then it's missing the SIP-010 trait methods and deployment will fail.

jferrant replied:
This is removed in my other PR. It is not required as far as I can tell.

jferrant commented on stacks-signer/src/main.rs:
```diff
@@ -169,6 +209,8 @@ fn main() {
     };
 
+    let (_cmd_send, cmd_recv) = channel();
+    let (res_send, _res_recv) = channel();
```
You should update the DKG and SIGN CLI commands to use this res_recv so that they terminate upon completion.

jferrant reviewed (changes requested):
I updated the state machine substantially and fixed the test to actually tear down the signers at the end. My initial check on the meaty portion (following node and signer setup), it takes ~15 seconds to run DKG and Sign a message.

jferrant committed:
Cleanup command and state machine within signer runloop

Signed-off-by: Jacinta Ferrant <jacinta@trustmachines.co>

jferrant committed:
Stop signers in test

Signed-off-by: Jacinta Ferrant <jacinta@trustmachines.co>

xoloki committed:
don't need to clone OperationResult, Signature, or SchnorrProof

xoloki commented:
I fixed the build so we don't expect any additional types to `impl Clone`, and rebased on the merged `signer-first` PR.  This PR should be ready to go now.

jcnelson commented on stacks-signer/src/main.rs:
```diff
@@ -79,22 +85,66 @@ fn write_chunk_to_stdout(chunk_opt: Option<Vec<u8>>) {
+
+// Process a DKG result
+fn process_dkg_result(dkg_res: &[OperationResult]) {
+    if dkg_res.len() != 1 {
```
Why doesn't this method just take a single `&OperationResult` instead of a slice reference?

jcnelson replied:
Same question for `process_sign_result()` below

jferrant replied:
Because the run one pass can in theory return multiple items (it doesn't right now because of how we do things sequentially).  I added this check because we would not expect more than one DKG result or more than one sign. However, handling it internally to this function call made more sense to me/removed the most amount of duplicate code. In theory of course I can assert the size of the list outside of this function call and access the first element in the list, but then we will have more duplicate code in handle_dkg, handle_sign, and handle_dkg_sign.

jcnelson replied:
I think it's fine as long as it's just part of `main`'s CLI handling.  I was mainly curious as to why the runloop itself doesn't just send a single `OperationResult` back instead of `Vec<OperationResult>`, because it seemed to be the case that the code always expected a `Vec` with exactly one element.  But if in the future there will be multiple items, then this is fine.

jcnelson commented on stacks-signer/src/runloop.rs:
```diff
@@ -68,60 +50,47 @@ pub struct RunLoop<C> {
+                    match command {
+                        RunLoopCommand::Dkg => {
+                            info!("Starting DKG");
+                            if let Ok(msg) = self.coordinator.start_distributed_key_generation() {
```
What happens if this fails? What's the `Err(..)` case?  Same question for `RunLoopCommand::Sign` below with `start_signing_message()`.

jferrant replied:
Good point. If it fails, we should not update the state to State::Dkg because it will become stuck. We should move the state update internal if let Ok() check. Currently if start_distributed_key_generation fails. Beyond this, it would imply that there is something substantially wrong with the coordinator state. Not sure if we should panic here, just warn. Attempt to push back the command and reprocess it? Not sure how it should fail here. @xoloki

jferrant replied:
Done :)

jcnelson commented on stacks-signer/src/runloop.rs:
```diff
@@ -68,60 +50,47 @@ pub struct RunLoop<C> {
+            State::Dkg | State::Sign => {
+                // We cannot execute the next command until the current one is finished...
+                // Do nothing
+                debug!("Waiting for operation to finish");
```
Does this command just get dropped then?  Asking because the signer daemon will receive a list of things it needs to sign (sBTC transactions, block headers), and it can't just drop them.  Will these requests be buffered up somewhere?

jferrant replied:
It gets stored in the command queue until it is able to process the next one. This will have to be reviewed down the line to enable processing multiple things at once, but currently it does it sequentially, buffering commands as they are received until it can continue processing them.

jferrant committed:
CRC to retry commands on failure  and clippy cleanup

Signed-off-by: Jacinta Ferrant <jacinta@trustmachines.co>

jferrant committed:
CRC: assert the length of the sign and dkg results instead of warning

Signed-off-by: Jacinta Ferrant <jacinta@trustmachines.co>

jferrant committed:
Clippy fixes for main.rs

Signed-off-by: Jacinta Ferrant <jacinta@trustmachines.co>

jcnelson reviewed (approved):
LGTM. Thanks for addressing my comments!

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
